### PR TITLE
[Android] ScrollView with TextInput not detecting Scroll Gesture

### PR DESCRIPTION
**Pull request by fabriziobertoglio1987**
## Summary

This issue fixes https://github.com/facebook/react-native/issues/25594 fixes https://github.com/facebook/react-native/issues/12167 fixes https://github.com/facebook/react-native/issues/13997
The ScrollView functionality is not compatible with TextInput scrolling. This pull request implements the scroll functionality of a TextInput in a ScrollView and fixes a bug causing ScrollView to not detect scroll gesture when including a TextInput with textAlign center. 

ADDITIONAL INFORMATION
canScrollHorizontally(-1) detects if a TextInput can scroll to the right, but it does not work with Gravity.CENTER. Consequence is that the TextInput can not be scrolled right and the parent component (in our example a ScrollView) is instead scrolled. For this reason the TextInput Gravity is set to Gravity.LEFT when focused for editing and set back to Gravity.CENTER once the focus leaves the TextInput. Similar issues with scrolling left (the TextInput never reaches the end).

This commit uses event onScrollChanged to set the gravity back to Gravity.CENTER if previously resetted. The gravity is set back to Gravity.CENTER once the focus leaves the TextInput and the Scrolling is ended.

The functionality allows the user to Scroll Both the TextInput and the Parent once the TextInput reached his end, yet other
functionality make the TextInput always move to his cursor position, which means that the cusor should move to scroll the TextInput, but changes to this that should be discussed in another pull request.

## Changelog

[Android] [Fixed] - ScrollView with TextInput not detecting Scroll Gesture and issues with textAlign center

## Test Plan

**<details><summary>CLICK TO OPEN TESTS RESULTS</summary>**
<p>

This pull request solves problems connected to https://github.com/facebook/react-native/commit/372d001a5d5e24b68d194e01c356d973642fac86 as explained in https://github.com/facebook/react-native/pull/25749#discussion_r433079311 and https://github.com/facebook/react-native/pull/25749#discussion_r433268959

The below example is available in RNTester. 

| **AFTER** | **AFTER** | 
|:-------------------------:|:-------------------------:|
|  |  |

</p>
</details>

### CODE REVIEW

**Comment by JoshuaGross** [comment]
@fabriziobertoglio1987 This doesn't work quite as I would expect. Specifically, in the second "horizontal" example, I can scroll the TextInput left but it seems there's no way to scroll the TextInput right, even if my finger is touching the TextInput and not the ScrollView space. That doesn't feel intuitively correct to me?

**Comment by JoshuaGross** [comment]
Maybe there's a different repro that would help me understand the issue. Using the example you provided, it seems correct without any Java changes and doesn't feel right with these Java changes.

**Comment by fabriziobertoglio1987** [comment]
Thanks a lot and Sorry @JoshuaGross for my mistakes.

The changes to fix the current Java functionalities are included in commits https://github.com/facebook/react-native/pull/29025/commits/de24234672f1ba2396abff28a2d2dc6e75969536 https://github.com/facebook/react-native/pull/29025/commits/514da1356cd2aad2530904c06ae9fc955b186711 and https://github.com/facebook/react-native/pull/29025/commits/5a8bf3a484206826278d92b632fdbb6e829c5d4a.
The commits messages include additional explanation of the changes.

I also improved RNTester to be easy to test with a Real Device. I tested with two devices and did not experience anymore issues.

Thanks a lot for your help and tolerance for my errors.
Sorry
I remain available for additional improvements
I wish you a good day
Fabrizio

| **HORIZONTAL** | **AFTER** | 
|:-------------------------:|:-------------------------:|
|  |  |

**Comment by JoshuaGross** [comment]
@fabriziobertoglio1987 No need to apologize, I appreciate you spending time on this.

I think the onscroll code makes sense now. My only remaining concern is with the scroll gravity logic - is it necessary? I think it could break some TextInputs if the default gravity for that TextInput isn't specified as LEFT. It seems to work fine without the gravity bit at all?

**Comment by fabriziobertoglio1987** [comment]
@JoshuaGross thanks a lot for the reviewing this pull request. I appreciate your feedback.

**QUESTION 1**
>I think the onscroll code makes sense now. My only remaining concern is with the scroll gravity logic - is it necessary?

The previous functionality was built in https://github.com/facebook/react-native/commit/372d001a5d5e24b68d194e01c356d973642fac86, @andreicoman11 may also have an interesting opinion on this change.

As demonstrated in the comment below from https://github.com/facebook/react-native/pull/25749#discussion_r433079311, commits https://github.com/facebook/react-native/pull/29025/commits/de24234672f1ba2396abff28a2d2dc6e75969536, https://github.com/facebook/react-native/pull/29025/commits/514da1356cd2aad2530904c06ae9fc955b186711, summary https://github.com/facebook/react-native/pull/29025#issue-426146706, [**canScrollHorizontally**](https://developer.android.com/reference/android/view/View#canScrollHorizontally(int)) does not work with `Gravity.CENTER` (scrolling left and right). Set `Gravity.Center` to `Gravity.LEFT`  fixes issues https://github.com/facebook/react-native/issues/25594, https://github.com/facebook/react-native/issues/12167 and https://github.com/facebook/react-native/issues/13997. 

The code below does not work with `Gravity.CENTER`:

https://github.com/facebook/react-native/blob/cfa42605989eee5a9de42bdb1259fb7f4d9451fb/ReactAndroid/src/main/java/com/facebook/react/views/textinput/ReactEditText.java#L183-L184
> The issue is caused by `canScrollHorizontally` returns allways `true` for:
> 
> ```js
> <TextInput multiline={false} textAlign="center" />
> ```
> 
> 
> 
> The method definition of [`canScrollHorizontally`](https://github.com/aosp-mirror/platform_frameworks_base/blob/e80b45506501815061b079dcb10bf87443bd385d/core/java/com/android/internal/widget/ViewPager.java#L2380-L2394).

**QUESTION 2**
>I think it could break some TextInputs if the default gravity for that TextInput isn't specified as LEFT. It seems to work fine without the gravity bit at all?

I am not currently aware of those scenarios, but I will keep researching. I did not invest time searching for an alternative solution to https://github.com/facebook/react-native/pull/29025/commits/514da1356cd2aad2530904c06ae9fc955b186711. Commit https://github.com/facebook/react-native/pull/29025/commits/514da1356cd2aad2530904c06ae9fc955b186711 fixes https://github.com/facebook/react-native/issues/25594, https://github.com/facebook/react-native/issues/12167 and https://github.com/facebook/react-native/issues/13997

I will be contributing to ReactNative master full-time until end of August 2020 and I will solve any related issue to this pull request in 1-2 days. I am subscribed to all react-native threads and I read all `Android` related issues or comments via email. I also reply to all messages if tagged as I have specific email filters that notify me. I am also willing to write test coverage for this functionality, if you are considering merging the pull request.

>I think it could break some TextInputs if the default gravity for that TextInput isn't specified as LEFT

I believe [the workaround](https://github.com/facebook/react-native/blob/3c41d8738ffc475255b4d713ab17fe5f9c5a22b0/ReactAndroid/src/main/java/com/facebook/react/views/textinput/ReactEditText.java#L343-L352) will only change the `Gravity` of the `TextInput` with `Gravity.CENTER` to `Gravity.LEFT` onFocus and set it back to `Gravity.CENTER` once the TextInput is not focused anymore.

```java
  @Override
  protected void onFocusChanged(boolean focused, int direction, Rect previouslyFocusedRect) {
    super.onFocusChanged(focused, direction, previouslyFocusedRect);
    if (focused && mSelectionWatcher != null) {
      mSelectionWatcher.onSelectionChanged(getSelectionStart(), getSelectionEnd());
    }
    boolean resetGravity = getGravity() == Gravity.CENTER &&
      mMultiLine == false;
    if(focused && resetGravity) {
      setGravity(Gravity.LEFT);
      mGravityReset = true;
    };
    if(!focused && mGravityReset) {
      setGravity(Gravity.CENTER);
      mGravityReset = false;
    };
```

I remain availalbe for further help and additional changes.
Thanks a lot
I wish you a good day
Fabrizio Bertoglio

File: ReactAndroid/src/main/java/com/facebook/react/views/textinput/ReactEditText.java [hunk 1]
```diff
@@ -361,6 +379,16 @@ protected void onFocusChanged(boolean focused, int direction, Rect previouslyFoc
     if (focused && mSelectionWatcher != null) {
       mSelectionWatcher.onSelectionChanged(getSelectionStart(), getSelectionEnd());
     }
+    boolean resetGravity = getGravity() == Gravity.CENTER &&
+      mMultiLine == false;
+    if(focused && resetGravity) {
+      setGravity(Gravity.LEFT);
+      mGravityReset = true;
+    };
```
[the remark this answers is not in the document]
Comment (fabriziobertoglio1987) [reply ReactAndroid/src/main/java/com/facebook/react/views/textinput/ReactEditText.java:386]: I have the same issue with [google-java-format 1.7](https://github.com/google/google-java-format)

**Pull request closed without merging**